### PR TITLE
iSentek ist8310 rewrite

### PR DESCRIPTION
**Pull request by dagar**
This is a cleanup/rewrite of the iSentek ist8310 driver in the style of the new IMU drivers.

 - simple state machine to reset, configure, etc (all sleeps removed)
 - checked register mechanism
 - if you disconnect and reconnect the sensor it will detect the bad settings and reconfigure itself
 - configured in 16 bit mode (1320 LSB/Gauss instead of 330 LSB/Gauss)

The other thing to note here is that I'm changing the rotation convention to something that makes more sense (consistent with the other new drivers) and finally killing the copied "convention" from the original 3dr compass/gps.

Screen Shot 2020-03-12 at 10 20 56 PM

The old driver kept z as is (down is up?) and swapped x & y.
The "new" convention I'm going with is to always respect what's obviously up (-z in this case), keep X as an anchor, and then flip Y as necessary to maintain the right hand rule. https://github.com/PX4/Firmware/pull/14184#issuecomment-588273190

The downside is having to update the rotations everywhere. The upside is it's quite a bit more intuitive to set it properly per board.

### CODE REVIEW

**Comment by dagar** [comment]
### Master
Screen Shot 2020-03-12 at 10 30 14 PM

### PR
Screen Shot 2020-03-12 at 10 29 14 PM

**Comment by dagar** [comment]
The driver rewrite itself isn't that interesting as the only change other than the structure is using the full 16 bit range. What's noteworthy here to be aware of is the handling of so called default orientations baked into the driver itself and our inability to properly set it everywhere. In particular the default IST8310 mounting in the Holybro or CUAV GPS/compass units is different than all the mRo units.

So my proposal is to remove all of that from the driver itself, and where possible we set the rotation in the board init. Then for the generic external ist8310 probe we set nothing. For the most part auto rotation during mag calibration (https://github.com/PX4/Firmware/pull/15120) should set it properly, but if not that fails at least it gives users the possibility of working out the proper rotation (eg CAL_MAGx_ROT) from the physical hardware.

Screenshot from 2020-08-25 14-15-45


Inevitably this will cause some pain for users with this hardware that already worked through this annoyance, however I'm more concerned about the users that didn't catch it until flight (or attempted flight I should say...). What's even more insidious is that you can slip through preflight mag inconsistency checks with multiple wrong externals.

Does anyone see a better solution?

**Comment by dagar** [comment]
Bench tested with a CUAV v5+, included GPS/compass module, and external mRo GPS/compass (ist8310 + lis3mdl). Here's what I get with everything mounted in the same direction.

``` Console
INFO  [commander] [cal] External Mag: 1 (396809), no rotation change: 0
INFO  [commander] [cal] External Mag: 2 (396833), determined rotation: 12
INFO  [commander] [cal] External Mag: 3 (335393), determined rotation: 2
INFO  [sensor_calibration] MAG 396825 EN: 1, offset: [-0.033 -0.010 -0.148], scale: [ 0.986  1.010  1.024], Internal
INFO  [sensor_calibration] MAG 396809 EN: 1, offset: [ 0.060  0.039 -0.103], scale: [ 0.991  1.003  1.011], External ROT: 0
INFO  [sensor_calibration] MAG 396833 EN: 1, offset: [-0.032  0.216 -0.244], scale: [ 1.027  0.970  1.008], External ROT: 12
INFO  [sensor_calibration] MAG 335393 EN: 1, offset: [-0.139 -0.178  0.013], scale: [ 0.999  0.996  1.006], External ROT: 2
```

``` Console
nsh> listener sensor_mag

TOPIC: sensor_mag 4 instances

Instance 0:
 sensor_mag_s
        timestamp: 129967796  (0.052176 seconds ago)
        timestamp_sample: 129966875  (921 us before timestamp)
        device_id: 396825 (Type: 0x06, I2C:3 (0x0E)) 
        x: 0.0348
        y: 0.2197
        z: 0.2674
        temperature: nan
        error_count: 0
        is_external: False

Instance 1:
 sensor_mag_s
        timestamp: 129963989  (0.057172 seconds ago)
        timestamp_sample: 129963071  (918 us before timestamp)
        device_id: 396809 (Type: 0x06, I2C:1 (0x0E)) 
        x: 0.1378
        y: 0.2560
        z: 0.2916
        temperature: nan
        error_count: 0
        is_external: True

Instance 2:
 sensor_mag_s
        timestamp: 129959776  (0.063904 seconds ago)
        timestamp_sample: 129958857  (919 us before timestamp)
        device_id: 396833 (Type: 0x06, I2C:4 (0x0E)) 
        x: -0.0992
        y: 0.4265
        z: -0.6515
        temperature: nan
        error_count: 0
        is_external: True

Instance 3:
 sensor_mag_s
        timestamp: 129986714  (0.038204 seconds ago)
        timestamp_sample: 129985236  (1478 us before timestamp)
        device_id: 335393 (Type: 0x05, I2C:4 (0x1E)) 
        x: 0.0583
        y: -0.2357
        z: 0.4215
        temperature: 31.3750
        error_count: 0
        is_external: True
```

After calibration the data all seems consistent and the estimated heading is correct.

``` Console
INFO  [vehicle_magnetometer] selected magnetometer: 396809 (1)
INFO  [data_validator] validator: best: 1, prev best: 1, failsafe: NO (0 events)
INFO  [data_validator] sensor #0, prio: 50, state: OK
INFO  [data_validator]  val:   0.0709, lp:   0.0732 mean dev:  -0.0001 RMS:   0.0020 conf:   1.0000
INFO  [data_validator]  val:   0.1939, lp:   0.1953 mean dev:  -0.0000 RMS:   0.0010 conf:   1.0000
INFO  [data_validator]  val:   0.4020, lp:   0.4042 mean dev:   0.0000 RMS:   0.0032 conf:   1.0000
INFO  [data_validator] sensor #1, prio: 75, state: OK
INFO  [data_validator]  val:   0.0636, lp:   0.0650 mean dev:  -0.0000 RMS:   0.0021 conf:   1.0000
INFO  [data_validator]  val:   0.2113, lp:   0.2104 mean dev:  -0.0000 RMS:   0.0012 conf:   1.0000
INFO  [data_validator]  val:   0.3905, lp:   0.3943 mean dev:  -0.0001 RMS:   0.0034 conf:   1.0000
INFO  [data_validator] sensor #2, prio: 75, state: OK
INFO  [data_validator]  val:   0.0499, lp:   0.0529 mean dev:  -0.0000 RMS:   0.0023 conf:   1.0000
INFO  [data_validator]  val:   0.1928, lp:   0.1921 mean dev:  -0.0000 RMS:   0.0030 conf:   1.0000
INFO  [data_validator]  val:   0.4234, lp:   0.4157 mean dev:  -0.0000 RMS:   0.0079 conf:   1.0000
INFO  [data_validator] sensor #3, prio: 75, state: OK
INFO  [data_validator]  val:   0.0625, lp:   0.0586 mean dev:   0.0000 RMS:   0.0027 conf:   1.0000
INFO  [data_validator]  val:   0.1936, lp:   0.1931 mean dev:  -0.0000 RMS:   0.0028 conf:   1.0000
INFO  [data_validator]  val:   0.4151, lp:   0.4149 mean dev:   0.0000 RMS:   0.0044 conf:   1.0000
INFO  [sensor_calibration] MAG 396825 EN: 1, offset: [-0.033 -0.010 -0.148], scale: [ 0.986  1.010  1.024], Internal
INFO  [sensor_calibration] MAG 396809 EN: 1, offset: [ 0.060  0.039 -0.103], scale: [ 0.991  1.003  1.011], External ROT: 0
INFO  [sensor_calibration] MAG 396833 EN: 1, offset: [-0.032  0.216 -0.244], scale: [ 1.027  0.970  1.008], External ROT: 12
INFO  [sensor_calibration] MAG 335393 EN: 1, offset: [-0.139 -0.178  0.013], scale: [ 0.999  0.996  1.006], External ROT: 2
```

Note, the external bus numbering is extremely confusing. On a CUAV v5+ the I2C4 labelled port is `-b 1`.

**Comment by dagar** [comment]
I was going to manually start the default IST8310 GPS/compass on other Holybro and CUAV boards, but I've removed it until I have an idea what's going on with the I2C external bus numbering.

File: src/drivers/magnetometer/isentek/ist8310/IST8310.cpp [hunk 1]
```diff
@@ -0,0 +1,298 @@
...
+					// register check failed, force reset
+					perf_count(_bad_register_perf);
+					Reset();
+					return;
+				}
+			}
+
+			// initiate next measurement
+			_state = STATE::MEASURE;
+			ScheduleNow();
```
Comment (bkueng) [inline src/drivers/magnetometer/isentek/ist8310/IST8310.cpp:229]: Just writing the register would be a bit more efficient here.

**Review by bkueng — COMMENTED** [review]
I guess the change of rotation will be caught with a mag inconsistency for existing users with external mag?

File: src/drivers/magnetometer/isentek/ist8310/IST8310.cpp [hunk 2]
```diff
@@ -0,0 +1,298 @@
...
+				uint8_t DATAZL;
+				uint8_t DATAZH;
+			} buffer{};
+
+			bool success = false;
+			uint8_t cmd = static_cast<uint8_t>(Register::STAT1);
+
+			if (transfer(&cmd, 1, (uint8_t *)&buffer, sizeof(buffer)) == PX4_OK) {
+
+				if (buffer.STAT1 && STAT1_BIT::DRDY) {
```
[the remark this answers is not in the document]
Comment (dagar) [reply src/drivers/magnetometer/isentek/ist8310/IST8310.cpp]: Thanks, fixed.

File: src/drivers/magnetometer/isentek/ist8310/IST8310.cpp (hunk 1 above)
Comment (dagar) [reply src/drivers/magnetometer/isentek/ist8310/IST8310.cpp:229]: I suppose it's a bit much. Fixed.

**Comment by dagar** [comment]
> I guess the change of rotation will be caught with a mag inconsistency for existing users with external mag?

Yes, although I'm hoping we can manually correct all of the known "good" cases (eg pixhawk 4) and keep them working unmodified.

For the remaining cases (eg most mRo units post HMC5883 EOL) that would have only been working with manual intervention I'm hoping we can catch it and encourage re-calibration (now with auto rotation). If that fails at least now they've got a shot at setting it correctly by hand from actually reviewing the unit.

**Comment by ryanjAA** [comment]
I have and have seen a bunch of different ist8310 mags have an i2c address of 0xF instead of 0xE (the latter is what is in the codebase). To test it we've changed the code and recompiled. Might be worth either looking at both addresses for this mag or having a way to select it if possible. 

https://github.com/PX4/Firmware/blob/24125b3da48908e8ff5cb59cb57c21a7e12749b6/src/drivers/magnetometer/isentek/ist8310/iSentek_IST8310_registers.hpp#L58

**Comment by dagar** [comment]
> I have and have seen a bunch of different ist8310 mags have an i2c address of 0xF instead of 0xE (the latter is what is in the codebase). To test it we've changed the code and recompiled. Might be worth either looking at both addresses for this mag or having a way to select it if possible.
> 
> https://github.com/PX4/Firmware/blob/24125b3da48908e8ff5cb59cb57c21a7e12749b6/src/drivers/magnetometer/isentek/ist8310/iSentek_IST8310_registers.hpp#L58

Thanks @ryanjAA, let's update it to probe for both. 

Can you point to any particular units that set it differently?

**Comment by ryanjAA** [comment]
Tested on master with alternate address. Doesn't exactly work (unless below functionality is what is intended):

Meaning when you run a compass calibration it doesn't pick it up.

If you run a ist8310 -X -a 15 start before the calibration it does but the moment you reboot the pixhawk (after the cal) it doesn't use or see it until you start it again manually so you could start it every time from the SD card at boot but i don't think that is the intention.

**Comment by ryanjAA** [comment]
Tested on master with alternate address. Doesn't exactly work (unless below functionality is what is intended):

Meaning when you run a compass calibration it doesn't pick it up.

If you run a ist8310 -X -a 15 start before the calibration it does but the moment you reboot the pixhawk (after the cal) it doesn't use or see it until you start it again manually so you could start it every time from the SD card at boot but i don't think that is the intention.

**Pull request merged**